lua/cfc_weapon_lockouts/server/sv_locker.lua

```diff
@@ -0,0 +1,216 @@
+end
+
+hook.Add( "PlayerSpawn", "CFC_WeaponLockouts_UnlockOnSpawn", function( ply )
+    ply.weaponLockouts = {}
+    ply.weaponLockoutWeapons = {}
+    lockWarns[ply] = {}
+
+    --Clear out warn info from weapons spawned by default
+    timer.Create( "CFC_WeaponLockouts_LockWarn_" .. ply:SteamID(), 0, 1, function()
+        lockWarns[ply] = {}
+    end )
+end )
+
+hook.Add( "PlayerDroppedWeapon", "CFC_WeaponLockouts_LockWeapon", function( ply, wep )
+    CFCWeaponLockouts.lockWeapon( ply, wep, true )
+end )
+
+hook.Add( "EntityRemoved", "CFC_WeaponLockouts_LockWeapon", function( ent )
+    if not IsValid( ent ) or not ent:IsWeapon() then return end
+
+    CFCWeaponLockouts.lockWeapon( ent )
+end )
+
+hook.Add( "PlayerSwitchWeapon", "CFC_WeaponLockouts_TrackWeaponOwner", function( ply, old, new )
+    if not IsValid( new ) then return end
+
+    new.weaponLockoutOwner = ply
+end )
+
+hook.Add( "WeaponEquip", "CFC_WeaponLockouts_CanPickup", function( wep, ply )
+    if not IsValid( wep ) or not IsValid( ply ) then return end
+
+    local weaponClass = wep:GetClass()
+    ply.weaponLockoutWeapons = ply.weaponLockoutWeapons or {}
+    wep.weaponLockoutOwner = ply
+
+    local plyWeapons = ply.weaponLockoutWeapons
+    local isLocked = CFCWeaponLockouts.weaponIsLocked( ply, weaponClass )
+
+    --In certain cases, EntityRemoved gets called after WeaponEquip, causing odd behavior where the weapon is unlocked, meant to be locked, and not held by the player.
+    --Manually keeping track of the weapon classes held by a player allows us to catch that error.
+    if not isLocked and plyWeapons[weaponClass] and not CFCWeaponLockouts.NOT_LOCKABLE[wep:GetClass()] then
+        isLocked = true
+        CFCWeaponLockouts.lockWeapon( ply, wep, false )
+    end
+
+    plyWeapons[weaponClass] = true
+
+    if isLocked then
+        local identifier = "CFC_WeaponLockouts_LockWarn_" .. ply:SteamID()
+        local warns = lockWarns[ply] or {}
+        warns[weaponClass] = "locked"
+        warns.lockedCount = ( warns.lockedCount or 0 ) + 1
+
+        timer.Create( identifier, CFCWeaponLockouts.WARN_BUILDUP, 1, function()
+            warnPlayer( identifier, ply, warns )
+        end )
+
+        local primaryAmmoType = wep:GetPrimaryAmmoType()
+        local secondaryAmmoType = wep:GetSecondaryAmmoType()
+        local primaryAmmo = ply:GetAmmoCount( primaryAmmoType ) or 0
+        local secondaryAmmo = ply:GetAmmoCount( secondaryAmmoType ) or 0
+        local clip1 = math.min( wep:Clip1() or 0, wep:GetMaxClip1() or 0 )
+        local clip2 = math.min( wep:Clip2() or 0, wep:GetMaxClip2() or 0 )
+
+        timer.Simple( 0, function()
+            if not IsValid( wep ) then return end
+            if primaryAmmo > 0 then ply:SetAmmo( 0, primaryAmmoType ) end
+            if secondaryAmmo > 0 then ply:SetAmmo( 0, secondaryAmmoType ) end
+            if clip1 > 0 then wep:SetClip1( 0 ) end
+            if clip2 > 0 then wep:SetClip2( 0 ) end
+        end )
```

Comment:
When setting gross timers like this, I like to add a comment explaining _why_. That way, in the future, if we know more about how this works, we won't be afraid to refactor the timer.